### ISSUE TITLE
Let TetGenIO read in subdomain IDs

### ISSUE THREAD

**Issue by YaqiWang** [libMesh/libmesh#1412]
Can we add an integer argument in `TetGenIO::read` to indicate that TetGen ele file contains an attribute which can be used to set the subdomain ID? `-1` means no subdomain ID, otherwise, we used it to access `element_attributes` and assign element subdomain ID. If you agree, I can send in a small PR on this.

**Comment by YaqiWang** [comment]
Oh, one more thing, although looks like we can use *.node to indicate a tetgen file based on line https://github.com/libMesh/libmesh/blob/master/src/mesh/namebased_io.C#L219, but actually the code will treat it as a nemsis file.

**Comment by jwpeterson** [comment]
> Oh, one more thing, although looks like we can use *.node to indicate a tetgen file based on line https://github.com/libMesh/libmesh/blob/master/src/mesh/namebased_io.C#L219, but actually the code will treat it as a nemsis file in previous lines

Are you talking about this logic?
```
if (name.rfind(".nem") + 4 == name.size() ||
      name.rfind(".n") + 2 == name.size())
```
I don't think either of these if-statements will trigger for a filename ending in ".node"

**Comment by jwpeterson** [comment]
NM, I found it on line 150. Yeah, that looks like a bug.

**Comment by YaqiWang** [comment]
Probably!

**Comment by YaqiWang** [comment]
I am thinking that the reader will need one more integer provided by us, which is not desirable. Let me read tetgen webpage more carefully to see if there is a sign in the file to indicate the existence of subdomain IDs.

**Comment by jwpeterson** [comment]
> NM, I found it on line 150. Yeah, that looks like a bug.

Nevermind again, I think the real bug may be in the `is_parallel_file_format()` function. The `name.rfind(".n")` check will cause the function to incorrectly return `true`.

```
inline bool
NameBasedIO::is_parallel_file_format (const std::string & name)
{
  return ((name.rfind(".xda") < name.size()) ||
          (name.rfind(".xdr") < name.size()) ||
          (name.rfind(".nem") < name.size()) ||
          (name.rfind(".n") < name.size())   ||
          (name.rfind(".cp") < name.size())
          );
}
```

**Comment by jwpeterson** [comment]
> Can we add an integer argument in TetGenIO::read to indicate that TetGen ele file contains an attribute which can be used to set the subdomain ID? -1 means no subdomain ID, otherwise, we used it to access element_attributes and assign element subdomain ID. If you agree, I can send in a small PR on this.

We already support reading element attributes from TetGen files (see `TetGenIO::element_in()`) if they are there. The `element_attributes` vector is a public member of the `TetGenIO` class, so I think the right approach in MOOSE is to wrap it (similar to what's done in e.g. Exodus.C) and use the `element_attributes` to assign subdomain ids.

Before you get too far down this path, however, I'd suggest making sure that TetGen is actually capable of producing the meshes you are after... I have had trouble with it entering infinite loops and failing in other ways in the past. Also, note that you will always have to build libmesh with `--disable-strict-lgpl` in order to use it, and you may not want to complicate life for all of your users if it's not really worth it.

**Comment by YaqiWang** [comment]
How about we assign the subdomain ID in place? Just above Line 236 of tetgen_io.C, we do
```
if (nAttri > 0)
   elem->subdomain_id() = element_attributes[j][0];
```

**Comment by YaqiWang** [comment]
TetGen will not be linked into the library. It is just used to generate .ele and .node files. Our reader does not call any tetgen routines. So license is not an issue.

**Comment by jwpeterson** [comment]
We can't assume this at the library level. What if the user is using the attributes to specify an elemental solution?

**Comment by YaqiWang** [comment]
On page http://wias-berlin.de/software/tetgen/1.5/doc/manual/manual006.html#ff_poly, it says:
```
If two values are written on a line after the x, y and z coordinate, the former is assumed to be a regional attribute (but will only be applied if the -A switch is selected), and the latter is assumed to be a regional volume constraint (but will only be applied if the -a switch is selected). It is possible to specify just one value after the coordinates. It can serve as both an attribute and a volume constraint, depending on the choice of switches. A negative maximum volume constraint allows to use the -A and the -a switches without imposing a volume constraint in this specific region.
```
And on http://wias-berlin.de/software/tetgen/1.5/doc/manual/manual005.html#cmd-A
```
The -A switch assigns an additional attribute (an **integer** number) to each tetrahedron that identifies to what facet-bounded region each tetrahedron belongs. In the output mesh, all tetrahedra in the same region will get a corresponding non-zero attribute.

- A region attribute is an **integer** which can be either **positive or negative**. **It must not be a zero**, which is used for the exterior of the PLC.
- B User-defined attributes are assigned to regions by the .poly or .smesh file (in the region section). If a region is **not** explicitly marked by the .poly file or .smesh file, tetrahedra in that region are **automatically** assigned a non-zero attribute.
-C By default, the region attributes are numbered from 1, 2, 3, ⋯. If there are user-assigned region attributes (by the .poly or .smesh file), the region attributes are shifted by a number i, i.e., i+1, i+2, i+3, ⋯, where i is either 0 or the maximum integer of the user-assigned region attributes.
-D The -A switch has an effect only if the -p switch is used and the -r switch is not.
```

**Comment by YaqiWang** [comment]
Looks like tetgen assume the attribute is an integer. I agree though it is not safe to assume it is the subdomain ID. I guess, we have to use a flag to indicate we want to do the assignment.

**Comment by jwpeterson** [comment]
> On page http://wias-berlin.de/software/tetgen/1.5/doc/manual/manual006.html#ff_poly, it says:

This doc is referring to the  syntax of a .poly file, we currently only support reading .ele files

If you scroll down to Section 5.2.4 on that same page about .ele files, it says 
```
An .ele file contains a list of tetrahedra.

  First line: <# of tetrahedra> <nodes per tet. (4 or 10)>
              <region attribute (0 or 1)>
  Remaining lines list # of tetrahedra:
    <tetrahedron #> <node> <node> ... <node> [attribute]
    ...
```
and
```
If the <region attribute> in the first line is 1, each tetrahedra has an additional region attribute
(an integer) in the last column. Region attributes of tetrahedra are tags used mainly to identify 
which tetrahedra of the tetrahedralization are associated with which facet-bounded region (sub-
domain) of the PLC, set in the fourth part of a .poly or a .smesh file. Region attributes do not diffuse
across facets, all tetrahedra in the same region have exactly the same region attribute. A common
application of the region attribute is to determine which material a tetrahedron has.
```

So it appears that you can only have a single attribute for each Tetrahedron anyway, and it is meant to be an integer that refers to something in a .poly file, but we could use it simply as a subdomain id. I'm not sure how we came up with our current approach to reading .ele files, perhaps it was based on an earlier Tetgen file format? But it sounds like we should actually update it to do what you want.

**Comment by jwpeterson** [comment]
> TetGen will not be linked into the library. It is just used to generate .ele and .node files. Our reader does not call any tetgen routines. So license is not an issue.

I see. Yes, that should be fine since the `TetGenIO` stuff is available regardless of the status of `--disable-strict-lgpl`.

**Comment by jwpeterson** [comment]
I'll make a PR for this and the `is_parallel_file_format()` today some time...

**Comment by jwpeterson** [comment]
The problems raised in this issue were addressed in #1413 and #1414.
